custom_components/octopus_intelligent/graphql_util.py

```diff
@@ -0,0 +1,63 @@
+"""Validation and parsing for the Octopus GraphQL API."""
+from ast import literal_eval
+from pprint import pformat
+
+from gql.transport.exceptions import TransportQueryError
+
+from .graphql_client import OctopusEnergyGraphQLClient
+
+
+class InvalidAuthError(Exception):
+    """Invalid Octopus API key or account number."""
+
+
+async def validate_octopus_account(client: OctopusEnergyGraphQLClient, account_id: str):
+    """Check that the Octopus account_id matches the authenticated API."""
+    try:
+        accounts = await client.async_get_accounts()
+    except TransportQueryError as ex:
+        msg = parse_gql_query_error(ex, "Authentication failed")
+        raise InvalidAuthError(msg) from ex
+    if account_id not in accounts:
+        raise InvalidAuthError(
+            f"Account '{account_id}' not found in accounts {accounts}"
+        )
```

Comment:
This function is called by both `OctopusIntelligentSystem.start()` and `config_flow.try_connection()`. I considered placing this function in the existing `util.py` module or in the existing `graphql_client.py` module. I thought it wasn’t a good fit for `util.py` because this function (and also `parse_gql_query_error()` below) has a dependency on `graphql_client.py` / graphql generally, and I didn’t think that such a dependency should be introduced in `util.py` (sooner or later, this new module or `graphql_client.py` might need to call a more basic function in `util.py`, and circular dependencies aren’t nice). And I thought it wasn’t a good fit for `graphql_client.py` because that module looks almost like it was auto-generated (was it?), from using a two-space indentation (unlike the rest of the repo) to having a rather terse style with functions having only one or two lines of code, starting with two underscore characters, or consisting of graphql schema strings. So I created this new module instead.

Reply:
The use of the proposed `parse_gql_query_error()` function that parses the “title” and “body” of GrahpQL responses provides informative log messages that distinguish between, for example, _“Too many requests”_ and _“Authentication failed”_ errors:

```
2024-03-06 16:53:08.614 DEBUG (MainThread) [custom_components.octopus_intelligent.config_flow] Trying to connect to Octopus during setup
2024-03-06 16:53:18.833 ERROR (MainThread) [custom_components.octopus_intelligent.config_flow] There were too many requests. Please try again later:
{'message': 'Too many requests.',
 'path': ['obtainKrakenToken'],
 'extensions': {'errorType': 'AUTHORIZATION',
                'errorCode': 'KT-CT-1199',
                'errorDescription': 'There were too many requests. Please try '
                                    'again later.’}}
```
```
2024-03-06 16:52:30.377 DEBUG (MainThread) [custom_components.octopus_intelligent.config_flow] Trying to connect to Octopus during setup
2024-03-06 16:52:39.609 ERROR (MainThread) [custom_components.octopus_intelligent.config_flow] Authentication failed:
{'message': 'Invalid data.',
 'path': ['obtainKrakenToken'],
 'extensions': {'errorType': 'VALIDATION',
                'errorCode': 'KT-CT-1139',
                'errorDescription': 'Authentication failed.',
                'errorClass': 'VALIDATION',
                'validationErrors': [{'message': 'Authentication failed.',
                                      'inputPath': ['input', 'apiKey']}]}}
```
That it looks a bit like a stack trace is a downside (it’s a dict, not a stack trace), but at least it is informative in case users copy and paste the logs in repo issues.